zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+// the constant isn't duplicated.
+const EPHEMERAL_SCOPE: TransparentKeyScope = match TransparentKeyScope::custom(2) {
+    Some(s) => s,
+    None => unreachable!(),
+};
+
+// Returns `TransparentAddressMetadata` in the ephemeral scope for the
+// given address index.
+pub(crate) fn metadata(address_index: NonHardenedChildIndex) -> TransparentAddressMetadata {
+    TransparentAddressMetadata::new(EPHEMERAL_SCOPE, address_index)
+}
+
+/// Returns the last reserved ephemeral address index in the given account,
+/// or -1 if the account has no reserved ephemeral addresses.
+pub(crate) fn last_reserved_index(
+    conn: &rusqlite::Connection,
+    account_id: AccountId,
+) -> Result<i32, SqliteClientError> {
+    match conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             WHERE account_id = :account_id
+             ORDER BY address_index DESC
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
+        )
+        .optional()?
+    {
+        Some(i) if i < 0 => Err(SqliteClientError::CorruptedData(
+            "negative index".to_owned(),
+        )),
+        Some(i) => Ok(i),
+        None => Ok(-1),
+    }
+}
+
+/// Returns the last ephemeral address index in the given account that
+/// would not violate the gap invariant if used.
+pub(crate) fn last_safe_index(
+    conn: &rusqlite::Connection,
+    account_id: AccountId,
+) -> Result<u32, SqliteClientError> {
+    // The inner join with `transactions` excludes addresses for which
+    // `mined_in_tx` is NULL. The query also excludes addresses observed
+    // to have been mined in a transaction that we currently see as unmined.
+    // This is conservative in terms of avoiding violation of the gap
+    // invariant: it can only cause us to get to the end of the gap sooner.
+    let last_mined_index: i32 = match conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             JOIN transactions t ON t.id_tx = mined_in_tx
+             WHERE account_id = :account_id AND t.mined_height IS NOT NULL
+             ORDER BY address_index DESC
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
+        )
+        .optional()?
+    {
+        Some(i) if i < 0 => Err(SqliteClientError::CorruptedData(
+            "negative index".to_owned(),
+        )),
+        Some(i) => Ok(i),
+        None => Ok(-1),
+    }?;
+    Ok(u32::try_from(last_mined_index.saturating_add(GAP_LIMIT)).unwrap())
+}
+
+/// Utility function to return an `InclusiveRange<u32>` that starts at `i + 1`
+/// and is of length up to `n`. The range is truncated if necessary to end at
+/// the maximum valid address index, `i32::MAX`.
+///
+/// Precondition: `i >= -1 and n > 0`
+pub(crate) fn range_after(i: i32, n: i32) -> RangeInclusive<u32> {
```

Comment:
Seems odd to use `i32` here, for similar reasons to those described above.

Reply:
I went through several iterations of the types, and this was by far the simplest way to eliminate potential overflows.

Reply:
The index types are now unsigned.